Implementation/47064 support custom field  in the attributesbytimestamp

dombesz opened the pull request:
https://community.openproject.org/wp/47064

dombesz committed:
[#47064] Support custom_field_* in the attributesByTimestamp

https://community.openproject.org/work_packages/47064

dombesz committed:
fixup! [#47064] Support custom_field_* in the attributesByTimestamp

dombesz committed:
fixup! [#47064] Support custom_field_* in the attributesByTimestamp

dombesz committed:
fixup! [#47064] Support custom_field_* in the attributesByTimestamp

dombesz committed:
fixup! [#47064] Support custom_field_* in the attributesByTimestamp

dombesz committed:
fixup! [#47064] Support custom_field_* in the attributesByTimestamp

dombesz committed:
Add comment about the attributes_by_timestamp method usage.

dombesz committed:
Add more specs and explanations

dombesz committed:
Add timestamp custom field specs to index_resource_spec

dombesz committed:
Load the custom_values for the singular #at_timestamp method

dombesz committed:
Add meaningful error message when the journalized.journal_id is not provided

dombesz committed:
Add specs for JournableDiffer edge case when value is changed from false to nil and vice versa.

ulferts commented on lib_static/plugins/acts_as_journalized/lib/acts/journalized/journable_differ.rb:
```diff
@@ -30,25 +30,29 @@ module Acts::Journalized
       def no_nil_to_empty_strings?(normalized_old_data, attribute, new_value)
         old_value = normalized_old_data[attribute]
-        new_value != old_value && (new_value.present? || old_value.present?)
+        new_value != old_value
```
I thinks this method no longer serves it's purpose of checking whether value might have changed from e.g. `nil` to `''` as the method name suggests. 

```
new_value = nil
old_value = ''

new_value != old_value && (new_value.present? || old_value.present?)
=> false

new_value != old_value
=> true
```

To have the implementation be more explicit, one could change it into
```
 ([new_value, old_value] - ['', nil]).any?
```

*Update note* In an earlier version of this `old_value = nil` was specified which was incorrect.

ulferts replied:
It is unsettling that no spec pics up on this. So I don't know how relevant my complaint is.

dombesz replied:
I made the changes here 953bdf0f2ddf3e6d9dcb06b40ebc919fbd3cebb1. The expectations now fail in case any of the `nil -> ''` attribute transitions are included.
Some minor adjustment was required on the conditions, because `[false].any?` returns `false` and it breaks the expectation when a transition happens from `nil -> false`. I used `([new_value, old_value] - ['', nil]).present?` instead.

dombesz committed:
Merge branch 'dev' into implementation/47064-support-custom_field_-in-the-attributesbytimestamp

ulferts commented on app/models/journable/with_historic_attributes/loader.rb:
```diff
@@ -86,6 +116,28 @@ def journalized_class
+      Journal
+        .where(id: journal_ids)
+        .includes({ customizable_journals: :custom_field })
+        .index_by(&:id)
```
Tiny thing: The `Journal` needn't be involved in the loading. The same result can be achieved by

```
Journal::CustomizableJournal
  .where(journal_id: journal_ids)
  .includes(:custom_field)
  .index_by(&:journal_id)
```

dombesz replied:
Done 57c3bf093da05728519db5c0e28f267524258281

ulferts reviewed (changes requested):
Hey Attila, I like the change. The number of places touched is really kept in check nicely.

I have two tiny remarks in the code and one issue that can also be addressed in a subsequent PR. That is actually what I would suggest, to fix the two small issues and have the bigger issue addressed separately.

The bigger issue is an N+1 query when the diff for the representer is created:

<img width="1440" alt="image" src="https://github.com/opf/openproject/assets/617519/b1e9f063-5d57-4051-892f-b70bb1a4e698">

We also have https://community.openproject.org/wp/48572 where we might place the fix in. It is not strictly the same but related.

dombesz committed:
Filter out the nil to empty string attribute changes in the JournableDiffer.

dombesz committed:
Load the custom_values association only on the Journable::WithHistoricAttributes

dombesz commented:
Thanks for the thorough review @ulferts. I addressed your remarks except the N+1 issue, that will be addressed separately. From my side this PR is good to go, once the CI is green.

ulferts reviewed (approved):
:+1: merge at will
